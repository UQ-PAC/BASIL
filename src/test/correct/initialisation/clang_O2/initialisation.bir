000005da: program
000005a8: sub __cxa_finalize(__cxa_finalize_result)
000005db: __cxa_finalize_result :: out u32 = low:32[R0]

000003a5:
0000049d: R16 := 0x20000
000004a4: R17 := mem[R16 + 8, el]:u64
000004aa: R16 := R16 + 8
000004af: call R17 with noreturn

000005a9: sub __do_global_dtors_aux(__do_global_dtors_aux_result)
000005dc: __do_global_dtors_aux_result :: out u32 = low:32[R0]

0000029b:
0000029f: #3 := R31 - 0x20
000002a5: mem := mem with [#3, el]:u64 <- R29
000002ab: mem := mem with [#3 + 8, el]:u64 <- R30
000002af: R31 := #3
000002b5: R29 := R31
000002bd: mem := mem with [R31 + 0x10, el]:u64 <- R19
000002c2: R19 := 0x20000
000002c9: R0 := pad:64[mem[R19 + 0x48]]
000002cf: when 0:0[R0] goto %000002cd
000005d0: goto %0000036e

0000036e:
00000371: R0 := 0x1F000
00000378: R0 := mem[R0 + 0xFC8, el]:u64
0000037e: when R0 = 0 goto %0000037c
000005d1: goto %00000395

00000395:
00000398: R0 := 0x20000
0000039f: R0 := mem[R0 + 0x28, el]:u64
000003a4: R30 := 0x730
000003a7: call @__cxa_finalize with return %0000037c

0000037c:
00000384: R30 := 0x734
00000386: call @deregister_tm_clones with return %00000388

00000388:
0000038b: R0 := 1
00000393: mem := mem with [R19 + 0x48] <- 7:0[R0]
000005d2: goto %000002cd

000002cd:
000002d7: R19 := mem[R31 + 0x10, el]:u64
000002de: R29 := mem[R31, el]:u64
000002e3: R30 := mem[R31 + 8, el]:u64
000002e7: R31 := R31 + 0x20
000002ec: call R30 with noreturn

000005ad: sub __libc_start_main(__libc_start_main_main, __libc_start_main_arg2, __libc_start_main_arg3, __libc_start_main_auxv, __libc_start_main_result)
000005dd: __libc_start_main_main :: in u64 = R0
000005de: __libc_start_main_arg2 :: in u32 = low:32[R1]
000005df: __libc_start_main_arg3 :: in out u64 = R2
000005e0: __libc_start_main_auxv :: in out u64 = R3
000005e1: __libc_start_main_result :: out u32 = low:32[R0]

000001f4:
00000487: R16 := 0x20000
0000048e: R17 := mem[R16, el]:u64
00000494: R16 := R16
00000499: call R17 with noreturn

000005ae: sub _fini(_fini_result)
000005e2: _fini_result :: out u32 = low:32[R0]

00000020:
00000026: #0 := R31 - 0x10
0000002c: mem := mem with [#0, el]:u64 <- R29
00000032: mem := mem with [#0 + 8, el]:u64 <- R30
00000036: R31 := #0
0000003c: R29 := R31
00000043: R29 := mem[R31, el]:u64
00000048: R30 := mem[R31 + 8, el]:u64
0000004c: R31 := R31 + 0x10
00000051: call R30 with noreturn

000005af: sub _init(_init_result)
000005e3: _init_result :: out u32 = low:32[R0]

000004f7:
000004fd: #5 := R31 - 0x10
00000503: mem := mem with [#5, el]:u64 <- R29
00000509: mem := mem with [#5 + 8, el]:u64 <- R30
0000050d: R31 := #5
00000513: R29 := R31
00000518: R30 := 0x5C8
0000051a: call @call_weak_fn with return %0000051c

0000051c:
00000521: R29 := mem[R31, el]:u64
00000526: R30 := mem[R31 + 8, el]:u64
0000052a: R31 := R31 + 0x10
0000052f: call R30 with noreturn

000005b0: sub _start(_start_result)
000005e4: _start_result :: out u32 = low:32[R0]

000001b5:
000001ba: R29 := 0
000001bf: R30 := 0
000001c5: R5 := R0
000001cc: R1 := mem[R31, el]:u64
000001d2: R2 := R31 + 8
000001d8: R6 := R31
000001dd: R0 := 0x1F000
000001e4: R0 := mem[R0 + 0xFD8, el]:u64
000001e9: R3 := 0
000001ee: R4 := 0
000001f3: R30 := 0x670
000001f6: call @__libc_start_main with return %000001f8

000001f8:
000001fb: R30 := 0x674
000001fe: call @abort with return %000005d3

000005d3:
000005d4: call @call_weak_fn with noreturn

000005b3: sub abort()


000001fc:
000004c9: R16 := 0x20000
000004d0: R17 := mem[R16 + 0x18, el]:u64
000004d6: R16 := R16 + 0x18
000004db: call R17 with noreturn

000005b4: sub call_weak_fn(call_weak_fn_result)
000005e5: call_weak_fn_result :: out u32 = low:32[R0]

00000200:
00000203: R0 := 0x1F000
0000020a: R0 := mem[R0 + 0xFD0, el]:u64
00000210: when R0 = 0 goto %0000020e
000005d5: goto %000003e5

0000020e:
00000216: call R30 with noreturn

000003e5:
000003e8: goto @__gmon_start__

000003e6:
000004b3: R16 := 0x20000
000004ba: R17 := mem[R16 + 0x10, el]:u64
000004c0: R16 := R16 + 0x10
000004c5: call R17 with noreturn

000005b6: sub deregister_tm_clones(deregister_tm_clones_result)
000005e6: deregister_tm_clones_result :: out u32 = low:32[R0]

0000021c:
0000021f: R0 := 0x20000
00000225: R0 := R0 + 0x48
0000022a: R1 := 0x20000
00000230: R1 := R1 + 0x48
00000236: #1 := ~R0
0000023b: #2 := R1 + ~R0
00000241: VF := extend:65[#2 + 1] <> extend:65[R1] + extend:65[#1] + 1
00000247: CF := pad:65[#2 + 1] <> pad:65[R1] + pad:65[#1] + 1
0000024b: ZF := #2 + 1 = 0
0000024f: NF := 63:63[#2 + 1]
00000255: when ZF goto %00000253
000005d6: goto %000003c7

000003c7:
000003ca: R1 := 0x1F000
000003d1: R1 := mem[R1 + 0xFC0, el]:u64
000003d6: when R1 = 0 goto %00000253
000005d7: goto %000003da

00000253:
0000025b: call R30 with noreturn

000003da:
000003de: R16 := R1
000003e3: call R16 with noreturn

000005b9: sub frame_dummy(frame_dummy_result)
000005e7: frame_dummy_result :: out u32 = low:32[R0]

000002f2:
000002f4: call @register_tm_clones with noreturn

000005ba: sub main(main_argc, main_argv, main_result)
000005e8: main_argc :: in u32 = low:32[R0]
000005e9: main_argv :: in out u64 = R1
000005ea: main_result :: out u32 = low:32[R0]

000002f6:
000002f9: R10 := 0x20000
000002ff: R10 := R10 + 0x40
00000304: R8 := 0x20000
00000309: R9 := 0x20000
0000030e: R12 := 0x20000
00000313: R0 := 0
0000031a: R11 := pad:64[mem[R10, el]:u32]
00000321: R13 := pad:64[mem[R8 + 0x30, el]:u32]
00000328: R14 := mem[R9 + 0x38, el]:u64
0000032e: R11 := pad:64[31:0[R11] + 3]
00000335: R15 := pad:64[mem[R12 + 0x34]]
0000033b: R13 := pad:64[31:0[R13] + 1]
00000341: R14 := R14 - 1
00000349: mem := mem with [R10 + 4, el]:u32 <- 31:0[R11]
0000034f: R10 := pad:64[31:0[R15] + 1]
00000357: mem := mem with [R8 + 0x30, el]:u32 <- 31:0[R13]
0000035f: mem := mem with [R9 + 0x38, el]:u64 <- R14
00000367: mem := mem with [R12 + 0x34] <- 7:0[R10]
0000036c: call R30 with noreturn

000005bb: sub register_tm_clones(register_tm_clones_result)
000005eb: register_tm_clones_result :: out u32 = low:32[R0]

0000025d:
00000260: R0 := 0x20000
00000266: R0 := R0 + 0x48
0000026b: R1 := 0x20000
00000271: R1 := R1 + 0x48
00000278: R1 := R1 + ~R0 + 1
0000027e: R2 := 0.63:63[R1]
00000285: R1 := R2 + (R1 ~>> 3)
0000028b: R1 := extend:64[63:1[R1]]
00000291: when R1 = 0 goto %0000028f
000005d8: goto %000003a9

000003a9:
000003ac: R2 := 0x1F000
000003b3: R2 := mem[R2 + 0xFE0, el]:u64
000003b8: when R2 = 0 goto %0000028f
000005d9: goto %000003bc

0000028f:
00000297: call R30 with noreturn

000003bc:
000003c0: R16 := R2
000003c5: call R16 with noreturn
