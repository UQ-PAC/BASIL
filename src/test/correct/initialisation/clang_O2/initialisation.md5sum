2afca183c8f64242b52d1ed2af7a7f1d  correct/initialisation/clang_O2/a.out
048ed5a5fa60c626abf5a47d0182b052  correct/initialisation/clang_O2/initialisation.adt
8e6ff5e1c4d7b4a8b1881d0500cb0b46  correct/initialisation/clang_O2/initialisation.bir
8bb45067e16a6b35e917dd5bac22d3fb  correct/initialisation/clang_O2/initialisation.relf
27f3459cf4f3690247b0c0ea44a9e5a4  correct/initialisation/clang_O2/initialisation.gts
