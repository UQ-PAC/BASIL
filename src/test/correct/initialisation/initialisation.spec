Globals:
x: bv32
y: bv8
z: bv64
a: bv32[2]

Subroutine: main
Ensures: x == 6bv32
Ensures: a[1] == 4bv32
Ensures: a[0] == 1bv32