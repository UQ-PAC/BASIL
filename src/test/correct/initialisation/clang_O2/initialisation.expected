var {:extern} Gamma_R0: bool;
var {:extern} Gamma_R10: bool;
var {:extern} Gamma_R11: bool;
var {:extern} Gamma_R12: bool;
var {:extern} Gamma_R13: bool;
var {:extern} Gamma_R14: bool;
var {:extern} Gamma_R15: bool;
var {:extern} Gamma_R8: bool;
var {:extern} Gamma_R9: bool;
var {:extern} Gamma_mem: [bv64]bool;
var {:extern} R0: bv64;
var {:extern} R10: bv64;
var {:extern} R11: bv64;
var {:extern} R12: bv64;
var {:extern} R13: bv64;
var {:extern} R14: bv64;
var {:extern} R15: bv64;
var {:extern} R8: bv64;
var {:extern} R9: bv64;
var {:extern} mem: [bv64]bv8;
const {:extern} $a_addr: bv64;
axiom ($a_addr == 69696bv64);
const {:extern} $x_addr: bv64;
axiom ($x_addr == 69680bv64);
const {:extern} $y_addr: bv64;
axiom ($y_addr == 69684bv64);
const {:extern} $z_addr: bv64;
axiom ($z_addr == 69688bv64);
function {:extern} L(mem$in: [bv64]bv8, index: bv64) returns (bool) {
  false
}

function {:extern} {:bvbuiltin "bvadd"} bvadd32(bv32, bv32) returns (bv32);
function {:extern} {:bvbuiltin "bvadd"} bvadd64(bv64, bv64) returns (bv64);
function {:extern} gamma_load32(gammaMap: [bv64]bool, index: bv64) returns (bool) {
  (gammaMap[bvadd64(index, 3bv64)] && (gammaMap[bvadd64(index, 2bv64)] && (gammaMap[bvadd64(index, 1bv64)] && gammaMap[index])))
}

function {:extern} gamma_load64(gammaMap: [bv64]bool, index: bv64) returns (bool) {
  (gammaMap[bvadd64(index, 7bv64)] && (gammaMap[bvadd64(index, 6bv64)] && (gammaMap[bvadd64(index, 5bv64)] && (gammaMap[bvadd64(index, 4bv64)] && (gammaMap[bvadd64(index, 3bv64)] && (gammaMap[bvadd64(index, 2bv64)] && (gammaMap[bvadd64(index, 1bv64)] && gammaMap[index])))))))
}

function {:extern} gamma_load8(gammaMap: [bv64]bool, index: bv64) returns (bool) {
  gammaMap[index]
}

function {:extern} gamma_store32(gammaMap: [bv64]bool, index: bv64, value: bool) returns ([bv64]bool) {
  gammaMap[index := value][bvadd64(index, 1bv64) := value][bvadd64(index, 2bv64) := value][bvadd64(index, 3bv64) := value]
}

function {:extern} gamma_store64(gammaMap: [bv64]bool, index: bv64, value: bool) returns ([bv64]bool) {
  gammaMap[index := value][bvadd64(index, 1bv64) := value][bvadd64(index, 2bv64) := value][bvadd64(index, 3bv64) := value][bvadd64(index, 4bv64) := value][bvadd64(index, 5bv64) := value][bvadd64(index, 6bv64) := value][bvadd64(index, 7bv64) := value]
}

function {:extern} gamma_store8(gammaMap: [bv64]bool, index: bv64, value: bool) returns ([bv64]bool) {
  gammaMap[index := value]
}

function {:extern} memory_load32_le(memory: [bv64]bv8, index: bv64) returns (bv32) {
  (memory[bvadd64(index, 3bv64)] ++ (memory[bvadd64(index, 2bv64)] ++ (memory[bvadd64(index, 1bv64)] ++ memory[index])))
}

function {:extern} memory_load64_le(memory: [bv64]bv8, index: bv64) returns (bv64) {
  (memory[bvadd64(index, 7bv64)] ++ (memory[bvadd64(index, 6bv64)] ++ (memory[bvadd64(index, 5bv64)] ++ (memory[bvadd64(index, 4bv64)] ++ (memory[bvadd64(index, 3bv64)] ++ (memory[bvadd64(index, 2bv64)] ++ (memory[bvadd64(index, 1bv64)] ++ memory[index])))))))
}

function {:extern} memory_load8_le(memory: [bv64]bv8, index: bv64) returns (bv8) {
  memory[index]
}

function {:extern} memory_store32_le(memory: [bv64]bv8, index: bv64, value: bv32) returns ([bv64]bv8) {
  memory[index := value[8:0]][bvadd64(index, 1bv64) := value[16:8]][bvadd64(index, 2bv64) := value[24:16]][bvadd64(index, 3bv64) := value[32:24]]
}

function {:extern} memory_store64_le(memory: [bv64]bv8, index: bv64, value: bv64) returns ([bv64]bv8) {
  memory[index := value[8:0]][bvadd64(index, 1bv64) := value[16:8]][bvadd64(index, 2bv64) := value[24:16]][bvadd64(index, 3bv64) := value[32:24]][bvadd64(index, 4bv64) := value[40:32]][bvadd64(index, 5bv64) := value[48:40]][bvadd64(index, 6bv64) := value[56:48]][bvadd64(index, 7bv64) := value[64:56]]
}

function {:extern} memory_store8_le(memory: [bv64]bv8, index: bv64, value: bv8) returns ([bv64]bv8) {
  memory[index := value[8:0]]
}

function {:extern} {:bvbuiltin "zero_extend 32"} zero_extend32_32(bv32) returns (bv64);
function {:extern} {:bvbuiltin "zero_extend 56"} zero_extend56_8(bv8) returns (bv64);
procedure {:extern} rely();
  modifies Gamma_mem, mem;
  ensures (Gamma_mem == old(Gamma_mem));
  ensures (mem == old(mem));
  free ensures (memory_load32_le(mem, 1908bv64) == 131073bv32);
  free ensures (memory_load64_le(mem, 69064bv64) == 1808bv64);
  free ensures (memory_load64_le(mem, 69072bv64) == 1728bv64);
  free ensures (memory_load64_le(mem, 69592bv64) == 1812bv64);
  free ensures (memory_load64_le(mem, 69672bv64) == 69672bv64);

procedure {:extern} rely_transitive();
  modifies Gamma_mem, mem;
  ensures (Gamma_mem == old(Gamma_mem));
  ensures (mem == old(mem));

implementation {:extern} rely_transitive()
{
  call rely();
  call rely();
}

procedure {:extern} rely_reflexive();

procedure {:extern} guarantee_reflexive();
  modifies Gamma_mem, mem;

procedure main();
  modifies Gamma_R0, Gamma_R10, Gamma_R11, Gamma_R12, Gamma_R13, Gamma_R14, Gamma_R15, Gamma_R8, Gamma_R9, Gamma_mem, R0, R10, R11, R12, R13, R14, R15, R8, R9, mem;
  free requires (memory_load64_le(mem, 69664bv64) == 0bv64);
  free requires (memory_load64_le(mem, 69672bv64) == 69672bv64);
  free requires (memory_load64_le(mem, 69680bv64) == 416611827717bv64);
  free requires (memory_load64_le(mem, 69688bv64) == 68719476735bv64);
  free requires (memory_load64_le(mem, 69696bv64) == 8589934593bv64);
  free requires (memory_load32_le(mem, 1908bv64) == 131073bv32);
  free requires (memory_load64_le(mem, 69064bv64) == 1808bv64);
  free requires (memory_load64_le(mem, 69072bv64) == 1728bv64);
  free requires (memory_load64_le(mem, 69592bv64) == 1812bv64);
  free requires (memory_load64_le(mem, 69672bv64) == 69672bv64);
  ensures (memory_load32_le(mem, $x_addr) == 6bv32);
  ensures (memory_load32_le(mem, bvadd64($a_addr, 4bv64)) == 4bv32);
  ensures (memory_load32_le(mem, bvadd64($a_addr, 0bv64)) == 1bv32);
  free ensures (memory_load32_le(mem, 1908bv64) == 131073bv32);
  free ensures (memory_load64_le(mem, 69064bv64) == 1808bv64);
  free ensures (memory_load64_le(mem, 69072bv64) == 1728bv64);
  free ensures (memory_load64_le(mem, 69592bv64) == 1812bv64);
  free ensures (memory_load64_le(mem, 69672bv64) == 69672bv64);

implementation main()
{
  lmain:
    assume {:captureState "lmain"} true;
    R10, Gamma_R10 := 69632bv64, true;
    R10, Gamma_R10 := bvadd64(R10, 64bv64), Gamma_R10;
    R8, Gamma_R8 := 69632bv64, true;
    R9, Gamma_R9 := 69632bv64, true;
    R12, Gamma_R12 := 69632bv64, true;
    R0, Gamma_R0 := 0bv64, true;
    call rely();
    R11, Gamma_R11 := zero_extend32_32(memory_load32_le(mem, R10)), (gamma_load32(Gamma_mem, R10) || L(mem, R10));
    call rely();
    R13, Gamma_R13 := zero_extend32_32(memory_load32_le(mem, bvadd64(R8, 48bv64))), (gamma_load32(Gamma_mem, bvadd64(R8, 48bv64)) || L(mem, bvadd64(R8, 48bv64)));
    call rely();
    R14, Gamma_R14 := memory_load64_le(mem, bvadd64(R9, 56bv64)), (gamma_load64(Gamma_mem, bvadd64(R9, 56bv64)) || L(mem, bvadd64(R9, 56bv64)));
    R11, Gamma_R11 := zero_extend32_32(bvadd32(R11[32:0], 3bv32)), Gamma_R11;
    call rely();
    R15, Gamma_R15 := zero_extend56_8(memory_load8_le(mem, bvadd64(R12, 52bv64))), (gamma_load8(Gamma_mem, bvadd64(R12, 52bv64)) || L(mem, bvadd64(R12, 52bv64)));
    R13, Gamma_R13 := zero_extend32_32(bvadd32(R13[32:0], 1bv32)), Gamma_R13;
    R14, Gamma_R14 := bvadd64(R14, 18446744073709551615bv64), Gamma_R14;
    call rely();
    assert (L(mem, bvadd64(R10, 4bv64)) ==> Gamma_R11);
    mem, Gamma_mem := memory_store32_le(mem, bvadd64(R10, 4bv64), R11[32:0]), gamma_store32(Gamma_mem, bvadd64(R10, 4bv64), Gamma_R11);
    assume {:captureState "%00000349"} true;
    R10, Gamma_R10 := zero_extend32_32(bvadd32(R15[32:0], 1bv32)), Gamma_R15;
    call rely();
    assert (L(mem, bvadd64(R8, 48bv64)) ==> Gamma_R13);
    mem, Gamma_mem := memory_store32_le(mem, bvadd64(R8, 48bv64), R13[32:0]), gamma_store32(Gamma_mem, bvadd64(R8, 48bv64), Gamma_R13);
    assume {:captureState "%00000357"} true;
    call rely();
    assert (L(mem, bvadd64(R9, 56bv64)) ==> Gamma_R14);
    mem, Gamma_mem := memory_store64_le(mem, bvadd64(R9, 56bv64), R14), gamma_store64(Gamma_mem, bvadd64(R9, 56bv64), Gamma_R14);
    assume {:captureState "%0000035f"} true;
    call rely();
    assert (L(mem, bvadd64(R12, 52bv64)) ==> Gamma_R10);
    mem, Gamma_mem := memory_store8_le(mem, bvadd64(R12, 52bv64), R10[8:0]), gamma_store8(Gamma_mem, bvadd64(R12, 52bv64), Gamma_R10);
    assume {:captureState "%00000367"} true;
    goto main_basil_return;
  main_basil_return:
    assume {:captureState "main_basil_return"} true;
    return;
}

